tests/test_spawner.py

```diff
@@ -753,3 +753,73 @@ def test_image_version_supports_component_gateway(self, monkeypatch):
     assert spawner._validate_image_version_supports_component_gateway('weird-unexpected-version-124.3.v2.2020-02-15') is True
     assert spawner._validate_image_version_supports_component_gateway('1.3.weird-version-again') is True
+
+  def test_validate_proto(self, monkeypatch):
+    import yaml
+
+    def test_read_file(*args, **kwargs):
+      config_string = open('./tests/test_data/unknown_fields.yaml', 'r').read()
+      return config_string
+
+    fake_creds = AnonymousCredentials()
+    mock_dataproc_client = mock.create_autospec(ClusterControllerClient(credentials=fake_creds))
+    mock_gcs_client = mock.create_autospec(storage.Client(credentials=fake_creds, project='project'))
+    spawner = DataprocSpawner(hub=Hub(), dataproc=mock_dataproc_client, gcs=mock_gcs_client, user=MockUser(), _mock=True, gcs_notebooks=self.gcs_notebooks)
+
+    # Prevents a call to GCS. We return the local file instead.
+    monkeypatch.setattr(spawner, "read_gcs_file", test_read_file)
+
+    spawner.project = "test-project"
+    spawner.region = "us-east1"
+    spawner.zone = "us-east1-d"
+    spawner.env_str = "test-env-str"
+    spawner.args_str = "test-args-str"
+    spawner.user_options = {
+      'cluster_type': 'basic_uri.yaml',
+      'cluster_zone': 'us-east1-d'
+    }
+
+    cleaned_config = spawner.get_cluster_definition('')
+    warnings = dataprocspawner.spawner._validate_proto(cleaned_config, Cluster)
+
+    # Check that we had appropriate warning messages
+    assert len(warnings) == 7
+    expected_warnings = [
+       'Removing unknown/bad value BAD_ENUM_VALUE for field consume_reservation_type.',
+       "Removing unknown field unknown_field for class <class 'google.cloud.dataproc_v1beta2.types.clusters.NodeInitializationAction'>",
+       'Removing unknown/bad value UNKNOWN_COMPONENT_1 for field optional_components.',
+       'Removing unknown/bad value UNKNOWN_COMPONENT_2 for field optional_components.',
+       'Removing unknown/bad value UNKNOWN_COMPONENT_3 for field optional_components.',
+       "Removing unknown field unknown_field_config_level for class <class 'google.cloud.dataproc_v1beta2.types.clusters.ClusterConfig'>",
+       "Removing unknown field unknown_field_top_level for class <class 'google.cloud.dataproc_v1beta2.types.clusters.Cluster'>",
+    ]
+    for w in expected_warnings:
+      assert w in warnings, f'Expected message {w} in warnings {warnings}'
+
+    raw_config = spawner.get_cluster_definition('')
+    # Construct expected output
+    del raw_config['unknown_field_top_level']
+    del raw_config['config']['unknown_field_config_level']
+    del raw_config['config']['initialization_actions'][0]['unknown_field']
+    del raw_config['config']['gce_cluster_config']['reservation_affinity']['consume_reservation_type']
+    raw_config['config']['software_config']['optional_components'] = [
+        'JUPYTER', 'ZEPPELIN', 'ANACONDA', 'PRESTO']
+
+    # Coerce both of the outputs to proto so we can easily compare equality
```

Comment:
Getting the following error: `TypeError: 'JUPYTER' has type str, but expected one of: int, long`
iirc, the raw_config might need to use Component['JUPYTER'].value , Component['ANACONDA'].value , etc...

Reply:
Tests work with proto-plus>=1.10.0